File: python/modele/reservable_item.py
from abc import ABC, abstractmethod
from datetime import datetime

class ReservableItem(ABC):
    def __init__(self, name):
        self.name = name
        self.reservations = []

    @abstractmethod
    def display_details(self):
        pass

    def is_available(self, start_date, end_date):
        for res in self.reservations:
            if (start_date <= res['end'] and end_date >= res['start']):
                return False
        return True

    def add_reservation(self, reservation):
        self.reservations.append(reservation)

File: typescript/modele/reservableItem.ts
import Reservation from "./reservation";

export default class ReservableItem {
  name: string;
  reservations: Reservation[] = [];

  constructor(name: string) {
    this.name = name;
  }

  isAvailable(start: Date, end: Date): boolean {
    return this.reservations.every(
      r => end <= r.start || start >= r.end
    );
  }

  addReservation(reservation: Reservation): void {
    this.reservations.push(reservation);
  }

  displayDetails(index: number): void {
    console.log(`${index + 1}. ${this.name}`);
  }
}
